### PR TITLE
Validate layout configs #4578

### LINKED ISSUE #4578: Validate layout configurations on load
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

As raised in this comment: https://github.com/sqlfluff/sqlfluff/pull/4558#discussion_r1142745101

At the moment, the layout configs are being validated _on use_ which is potentially flaky and convoluted. Better would be to validate configs _on load_.

### Use case

_No response_

### Dialect

all

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by alanmcruickshank**
This fixes #4578 . It also adds a couple of other nice features at the same time:
- Inline config values were previously not validated at all (particularly not against the "removed" config list). They now are, for both that and for layout values.
- Inline config values also would only work if they began with _exactly_ `-- sqlfluff` (i.e. explicitly with a space). They can now also begin with `--sqlfluff` (i.e. without a space). I've not changed the docs to _suggest_ this approach, because I think using the spaced approach is better, but this means that it doesn't fail silently in the other cases.

This also adds a few test cases to cover each of them.

### CODE REVIEW

File: src/sqlfluff/core/config.py [hunk 1]
```diff
@@ -549,12 +565,37 @@ def _validate_configs(
...
+                if (
+                    # Key length must be 4
+                    (len(k) != 4)
+                    # Second value must (currently) be "type"
+                    or (k[1] != "type")
+                    # Last key value must be one of the allowable options.
+                    or (k[3] not in ALLOWABLE_LAYOUT_CONFIG_KEYS)
+                ):
+                    raise SQLFluffUserError(
+                        f"Config file {file_path!r} set an in valid `layout` option "
```
Comment (greg-finley) [inline src/sqlfluff/core/config.py:593]:
```suggestion
                        f"Config file {file_path!r} set an invalid `layout` option "
```

File: src/sqlfluff/core/config.py [hunk 2]
```diff
@@ -549,12 +565,37 @@ def _validate_configs(
...
                         " for more details."
                     )
 
+            # Second validate any layout configs for validity.
+            # NOTE: For now we don't check that the "type" is a valid one
+            # to reference, or that the values are valid. For the values,
+            # these are likely to be rejected by the layout routines at
+            # runtime. The last risk area is validating that the type is
+            # a valid one.
+            if k[0] == "layout":
```
Comment (greg-finley) [inline src/sqlfluff/core/config.py:580]: Pardon the ignorant question, but do we need to check that k has a length?

```suggestion
            if k and k[0] == "layout":
```

Comment (alanmcruickshank) [reply src/sqlfluff/core/config.py:580]: We probably don't explicitly - but that does seem sensible 🤔

File: src/sqlfluff/core/config.py [hunk 3]
```diff
@@ -549,12 +565,37 @@ def _validate_configs(
...
+                # Check for:
+                # - Key length
+                # - Key values
+                if (
+                    # Key length must be 4
+                    (len(k) != 4)
+                    # Second value must (currently) be "type"
+                    or (k[1] != "type")
+                    # Last key value must be one of the allowable options.
+                    or (k[3] not in ALLOWABLE_LAYOUT_CONFIG_KEYS)
```
[the remark this answers is not in the document]
Comment (alanmcruickshank) [reply src/sqlfluff/core/config.py:590]: I don't think so - if `k` isn't exactly 4 long, I don't believe that this part will ever be executed. We make sure it's 4 long a few lines above. That scenario is also covered by the test cases, so I think it's safe.

**Pull request merged**